keps/sig-network/3866-nftables-proxy/README.md

```diff
@@ -0,0 +1,1490 @@
+a single "canary" chain, and check every 30 seconds that the canary
+had not been deleted, and only recreate everything from scratch if the
+canary disappears.
+
+NFTables provides a way to monitor for changes without doing polling;
+you can keep a netlink socket open to the kernel (or a pipe open to an
+`nft monitor` process) and receive notifications when particular kinds
+of nftables objects are created or destroyed.
+
+However, the "everyone uses their own table" design of nftables means
+that this may not be necessary. IPTables-based firewall
+implementations flush all iptables rules because everyone's iptables
+rules are all mixed together and it's hard to do otherwise. But in
+nftables, a firewall ought to only flush _its own_ table when
+restarting, and leave everyone else's tables untouched. In particular,
+firewalld works this way when using nftables. We will need to see what
+other firewall implementations do.
+
+### High level design
+
+At a high level, the new mode should have the same architecture as the
+existing modes; it will use the service/endpoint-tracking code in
+`k8s.io/kubernetes/pkg/proxy` (or its eventual replacement from kpng)
+to watch for changes, and update rules in the kernel accordingly.
+
+### Low level design
+
+Some details will be figured out as we implement it. We may start with
+an implementation that is architecturally closer to the `iptables`
+mode, and then rewrite it to take advantage of additional nftables
+features over time.
+
+#### Tables
+
+Unlike iptables, nftables does not have any reserved/default tables or
+chains (eg, `nat`, `PREROUTING`). Users are expected to create their
+own tables and chains for their own purposes. An nftables table can
+only contain rules for a single "family" (`ip` (v4), `ip6`, `inet`
+(both IPv4 and IPv6), `arp`, `bridge`, or `netdev`), but unlike in
+iptables, you can have both "filter"-type chains and "NAT"-type chains
+in the same table.
+
+So, we will create a single `kube_proxy` table in the `ip` family, and
+another in the `ip6` family. All of our chains, sets, maps, etc, will
+go into those tables. Other system components (eg, firewalld) should
+ignore our table, so we should not need to worry about watching for
+other people deleting our rules like we have to in the `iptables`
+backend.
+
+(In theory, instead of creating one table each in the `ip` and `ip6`
+families, we could create a single table in the `inet` family and put
+both IPv4 and IPv6 chains/rules there. However, this wouldn't really
+result in much simplification, because we would still need separate
+sets/maps to match IPv4 addresses and IPv6 addresses. (There is no
+data type that can store/match either an IPv4 address or an IPv6
+address.) Furthermore, because of how Kubernetes Services evolved in
+parallel with the existing kube-proxy implementation, we have ended up
+with a dual-stack Service semantics that is most easily implemented by
+handling IPv4 and IPv6 completely separately anyway.)
+
```

Comment:
`inet` might be useful if we want to make a `KUBE-NFTABLES-HINT` table.

Reply:
We won't need that; `KUBE-IPTABLES-HINT` is needed because iptables has two modes and containerized apps need some way to figure out which mode the host filesystem is configured to use. But there's no corresponding thing with nftables. (And, if there was, then the hint would have to be set by kubelet, not kube-proxy anyway.)

Reply:
It's also a hint about whether `kube-proxy` is using iptables or nftables. At least, I think it would be.

Reply:
It's not, though; `KUBE-IPTABLES-HINT` is created by kubelet, regardless of the kube-proxy mode (or even if you're using a third-party proxy implementation). It's not "HINT that KUBErnetes is using IPTABLES", it's "HINT provided by KUBErnetes on how to use IPTABLES".